src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1477,6 +1478,26 @@ public async Task WaitForExitAsync(CancellationToken cancellationToken = default
             finally
             {
                 Exited -= handler;
+
+                await WaitUntilOutputEOF().ConfigureAwait(false);
+            }
+
+            // Wait until output streams have been drained
+            async ValueTask WaitUntilOutputEOF()
+            {
+                // if we have a hard timeout, we cannot wait for the streams
```

Comment:
How does this comment apply to the subsequent check?

Reply:
The comment is copied over from the sync equivalents. I suppose cancellation tokens are used to encode timeouts in this context so though I'd preserve it.

Reply:
The problem is it's not doing that.  If we actually want to avoid waiting if it's cancelable, then the check needs to be for CanBeCanceled rather than IsCancellationRequested.

Reply:
Ah, good point. In that case we should probably seek to just replicate sync behaviour? Alternatively we could just pass the cancellation token on to the reader.

Reply:
I updated the logic to check for CanBeCanceled, which should match sync behaviour.